contracts/HarmonyLightClient.sol

```diff
@@ -137,7 +144,33 @@ contract HarmonyLightClient is
 
     }
 
+    function _deleteEpoch(uint256 epochNo) private {
+        uint256[] storage blockNumbers = epochCheckPointBlockNumbers[epochNo];
+        for(uint256 i = 0; i < blockNumbers.length; i++) {
+            uint256 blockNo = blockNumbers[i];
+            bytes32 mmrRoot = checkPointBlocks[blockNo].mmrRoot;
+            delete checkPointBlocks[blockNo];
+            delete epochMmrRoots[epochNo][mmrRoot];
+        }
+        delete epochCheckPointBlockNumbers[epochNo];
+    }
+ 
+    function _pruneEpochs(uint256 pruneTime) private {
+        uint256 epochCur = oldestEpochStored;
+        if (epochCheckPointBlockNumbers[epochCur].length == 0) return;
+        for(uint256 i = 0; i < MAX_PRUNE_ONCE; i++) {
+            uint256 epochNo = epochCur++;
+            uint256[] storage epochBlockNumbers = epochCheckPointBlockNumbers[epochCur];
+            uint256 epochBlockCount = epochBlockNumbers.length;
+            if (epochBlockCount == 0) break;
+            if (checkPointBlocks[epochBlockNumbers[epochBlockCount-1]].time > pruneTime) break;
+            _deleteEpoch(epochNo);
+        }
+        if(epochCur > oldestEpochStored) oldestEpochStored = epochCur;
+    }
+
     function submitCheckpoint(bytes memory rlpHeader) external onlyRelayers whenNotPaused {
+        _pruneEpochs(block.timestamp - BLOCK_EXPIRED);
```

Comment:
If relayers are to be decentralized or made permissionless later, this may face the same issue as commented above regarding EthereumLightClient - what if malicious relayers flood the contract with epochs on a forked chain and cause the contract to prune all existing valid epochs in storage?

Reply:
resolved in another thread